File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid/Presenter/SharedView/SharedView_PopupControlSelectUserGroupMulti.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.App;
using Android.Support.V7.Widget;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Class;
using BPMOPMobile.Droid.Class;
using BPMOPMobile.Droid.Core.Common;
using BPMOPMobile.Droid.Core.Controller;
using BPMOPMobile.Droid.Presenter.Adapter;
using BPMOPMobile.Droid.Presenter.Fragment;
using Com.Google.Android.Flexbox;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace BPMOPMobile.Droid.Presenter.SharedView
{
    class SharedView_PopupControlSelectUserGroupMulti : SharedView_PopupControlBase
    {
        private ControllerDetailWorkflow CTRLDetailWorkflow = new ControllerDetailWorkflow();
        private bool _isUserAndGroup { get; set; }
        public SharedView_PopupControlSelectUserGroupMulti(LayoutInflater _inflater, AppCompatActivity _mainAct, CustomBaseFragment _fragment, string _fragmentTag, View _rootView, bool _isUserAndGroup) : base(_inflater, _mainAct, _fragment, _fragmentTag, _rootView)
        {
            this._isUserAndGroup = _isUserAndGroup;
        }

        public override void InitializeValue_Master(ViewElement clickedElement)
        {
            base.InitializeValue_Master(clickedElement);
        }

        public override void InitializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JObject JObjectChild)
        {
            base.InitializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
        }

        public override void InitializeView()
        {
            base.InitializeView();
            try
            {
                #region Get View - Init Data
                View _viewPopupControl = _inflater.Inflate(Resource.Layout.PopupControl_ChooseUser, null);
                LinearLayout _lnChooseUser = _viewPopupControl.FindViewById<LinearLayout>(Resource.Id.ln_PopupControl_ChooseUser);
                LinearLayout _lnContent = _viewPopupControl.FindViewById<LinearLayout>(Resource.Id.ln_PopupControl_ChooseUser_Content);
                LinearLayout _lnSearch = _viewPopupControl.FindViewById<LinearLayout>(Resource.Id.ln_PopupControl_ChooseUser_Search);
                ImageView _imgCloseChooseUser = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_ChooseUser_Close);
                ImageView _imgAcceptChooseUser = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_ChooseUser_Accept);
                ImageView _imgDeleteChooseUser = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_ChooseUser_Delete);
                TextView _tvTitleChooseUser = _viewPopupControl.FindViewById<TextView>(Resource.Id.tv_PopupControl_ChooseUser_Title);
                EditText _edtSearchChooseUser = _viewPopupControl.FindViewById<EditText>(Resource.Id.edt_PopupControl_ChooseUser_Search);
                RecyclerView _recyChooseUser = _viewPopupControl.FindViewById<RecyclerView>(Resource.Id.recy_PopupControl_ChooseUser);
                CustomFlexBoxRecyclerView _recySelectedUser = _viewPopupControl.FindViewById<CustomFlexBoxRecyclerView>(Resource.Id.recy_PopupControl_SelectedUser);

                _recySelectedUser.Visibility = ViewStates.Visible;

                if (elementParent.Enable)
                {
                    _lnSearch.Visibility = ViewStates.Visible;
                    _imgDeleteChooseUser.Visibility = ViewStates.Visible;
                    _imgAcceptChooseUser.Visibility = ViewStates.Visible;
                    _recySelectedUser.SetMaxRowAndRowHeight((int)CmmDroidFunction.ConvertDpToPixel(35, _rootView.Context), 3); // 95 *3
                }
                else
                {
                    _lnSearch.Visibility = ViewStates.Gone;
                    _imgDeleteChooseUser.Visibility = ViewStates.Gone;
                    _imgAcceptChooseUser.Visibility = ViewStates.Invisible;

                    //trừ đi top navigation và / 35 số cứng
                    int maxlines = (_mainAct.Resources.DisplayMetrics.HeightPixels - (int)CmmDroidFunction.ConvertDpToPixel(50, _rootView.Context)) / (int)CmmDroidFunction.ConvertDpToPixel(35, _rootView.Context); //

                    _recySelectedUser.SetMaxRowAndRowHeight((int)CmmDroidFunction.ConvertDpToPixel(35, _rootView.Context), maxlines); // 95 *3
                }

                if (_isUserAndGroup == true)
                    _edtSearchChooseUser.Hint = CmmFunction.GetTitle("MESS_REQUIRE_USERGROUP", "Vui lòng chọn người hoặc nhóm để thực hiện");
                else
                    _edtSearchChooseUser.Hint = CmmFunction.GetTitle("MESS_REQUIRE_USER", "Vui lòng chọn người thực hiện");


                List<BeanUserAndGroup> _lstUserAndGroupAll = new List<BeanUserAndGroup>();
                List<BeanUserAndGroup> _lstSelected = new List<BeanUserAndGroup>();

                switch (base.flagView)
                {
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                    default:
                        {
                            _tvTitleChooseUser.Text = elementParent.Title;
                            if (!String.IsNullOrEmpty(elementParent.Value))
                            {
                                if (!String.IsNullOrEmpty(elementParent.Value))
                                    _lstSelected = JsonConvert.DeserializeObject<List<BeanUserAndGroup>>(elementParent.Value);
                            }

                            break;
                        }
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                        {
                            _tvTitleChooseUser.Text = elementPopup.Title;
                            if (!String.IsNullOrEmpty(elementPopup.Value))
                            {
                                if (!String.IsNullOrEmpty(elementPopup.Value))
                                    _lstSelected = JsonConvert.DeserializeObject<List<BeanUserAndGroup>>(elementPopup.Value);
                            }
                            break;
                        }
                }

                SQLite.SQLiteConnection conn = new SQLite.SQLiteConnection(CmmVariable.M_DataPath);

                _lstUserAndGroupAll = conn.Query<BeanUserAndGroup>(_isUserAndGroup == true
                                                                    ? CTRLDetailWorkflow._queryBeanUserGroup
                                                                    : CTRLDetailWorkflow._queryBeanUser);
                conn.Close();


                if (_lstUserAndGroupAll != null && _lstUserAndGroupAll.Count > 0 && _lstSelected != null && _lstSelected.Count > 0) // Người đã được chọn sẽ không hiển thị vào list     
                {
                    for (int i = 0; i < _lstSelected.Count; i++)
                    {
                        _lstUserAndGroupAll = _lstUserAndGroupAll.Where(x => !x.ID.Equals(_lstSelected[i].ID)).ToList();
                    }
                }

                AdapterSelectUserGroupMultiple _adapterListUser = new AdapterSelectUserGroupMultiple((MainActivity)base._mainAct, _rootView.Context, _lstUserAndGroupAll, _lstSelected);
                AdapterSelectUserGroupMultiple_Text _adapterListUserSelected = new AdapterSelectUserGroupMultiple_Text((MainActivity)base._mainAct, _rootView.Context, _lstSelected, elementParent.Enable);
                FlexboxLayoutManager flexboxLayoutManager = new FlexboxLayoutManager(_rootView.Context);
                flexboxLayoutManager.FlexDirection = FlexDirection.Row;
                flexboxLayoutManager.JustifyContent = JustifyContent.FlexStart;

                _recySelectedUser.SetAdapter(_adapterListUserSelected);
                _recySelectedUser.SetLayoutManager(flexboxLayoutManager);
                #endregion

                #region Show View                
                Dialog _dialogPopupControl = new Dialog(_rootView.Context, Resource.Style.Theme_Custom_BPMOP_Dialog_FullScreen);
                Window window = _dialogPopupControl.Window;
                _dialogPopupControl.RequestWindowFeature(1);
                _dialogPopupControl.SetCanceledOnTouchOutside(false);
                _dialogPopupControl.SetCancelable(true);
                window.SetSoftInputMode(SoftInput.AdjustPan); // Để Hide soft keyboard ko bị Resize
                window.SetGravity(GravityFlags.Bottom);
                var dm = _mainAct.Resources.DisplayMetrics;

                _dialogPopupControl.SetContentView(_viewPopupControl);
                _dialogPopupControl.Show();
                WindowManagerLayoutParams s = window.Attributes;
                s.Width = WindowManagerLayoutParams.MatchParent;
                s.Height = WindowManagerLayoutParams.MatchParent;
                window.Attributes = s;
                #endregion

                #region Event
                if (elementParent.Enable == true) // Enabled = true mới cho clik
                {


                    _edtSearchChooseUser.TextChanged += (sender, e) =>
                    {
                        List<BeanUserAndGroup> _lstSearch = new List<BeanUserAndGroup>();
                        if (!String.IsNullOrEmpty(_edtSearchChooseUser.Text))
                        {
                            _lstSearch = _lstUserAndGroupAll.Where(x => x.Email != null).ToList();
                            _lstSearch = _lstSearch.Where(x => CmmFunction.removeSignVietnamese(x.Name).ToLowerInvariant().Contains(CmmFunction.removeSignVietnamese(_edtSearchChooseUser.Text).ToLowerInvariant())
                                                            || CmmFunction.removeSignVietnamese(x.Email).ToLowerInvariant().Contains(CmmFunction.removeSignVietnamese(_edtSearchChooseUser.Text).ToLowerInvariant())).ToList();
                        }
                        else
                        {
                            //_lstSearch = _lstUserAndGroupAll.ToList();
                            _lstSearch = new List<BeanUserAndGroup>();
                        }

                        _adapterListUser = new AdapterSelectUserGroupMultiple((MainActivity)base._mainAct, _rootView.Context, _lstSearch, _lstSelected);
                        StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(1, LinearLayoutManager.Vertical);

                        _recyChooseUser.SetAdapter(_adapterListUser);
                        _recyChooseUser.SetLayoutManager(staggeredGridLayoutManager);

                        _adapterListUser.CustomItemClick += (sender, e) =>
                        {
                            BeanUserAndGroup _clickedItem = e;
                            if (_clickedItem != null)
                            {
                                _lstSelected.Add(e);
                                _lstUserAndGroupAll = _lstUserAndGroupAll.Where(x => !x.ID.Equals(_clickedItem.ID)).ToList();

                                _adapterListUser.UpdateCurrentList(_lstUserAndGroupAll);
                                _adapterListUser.NotifyDataSetChanged();
                                _adapterListUserSelected.UpdateItemListIsClicked(_lstSelected);
                                _adapterListUserSelected.NotifyDataSetChanged();
                                _edtSearchChooseUser.Text = _edtSearchChooseUser.Text; // để set Adapter lại
                                _edtSearchChooseUser.SetSelection(_edtSearchChooseUser.Text.Length);

                                _recySelectedUser.SmoothScrollToPosition(_lstSelected.Count); // focus lại vi trí cuối cùng
                            }
                            _edtSearchChooseUser.Text = "";
                        };
                        _adapterListUserSelected.CustomItemClick += (sender, e) =>
                        {
                            BeanUserAndGroup _clickedItem = e;
                            if (_clickedItem != null)
                            {
                                _lstUserAndGroupAll.Add(e);
                                _lstSelected = _lstSelected.Where(x => !x.ID.Equals(_clickedItem.ID)).ToList(); // Loại Item vừa click ra

                                _adapterListUser.UpdateCurrentList(_lstUserAndGroupAll);
                                _adapterListUser.NotifyDataSetChanged();
                                _adapterListUserSelected.UpdateItemListIsClicked(_lstSelected);
                                _adapterListUserSelected.NotifyDataSetChanged();
                                _edtSearchChooseUser.Text = _edtSearchChooseUser.Text; // để set Adapter lại
                                _edtSearchChooseUser.SetSelection(_edtSearchChooseUser.Text.Length);
                            }
                        };
                    };
                }

                _imgCloseChooseUser.Click += (sender, e) =>
                {
                    CmmDroidFunction.HideSoftKeyBoard(_edtSearchChooseUser, _mainAct);
                    _dialogPopupControl.Dismiss();
                };

                _imgDeleteChooseUser.Click += (sender, e) =>
                {
                    CmmDroidFunction.HideSoftKeyBoard(_edtSearchChooseUser, _mainAct);
                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, "");
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, "");
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                    }
                };

                _imgAcceptChooseUser.Click += (sender, e) =>
                {
                    CmmDroidFunction.HideSoftKeyBoard(_edtSearchChooseUser, _mainAct);
                    List<BeanUserAndGroup> _lstResult = new List<BeanUserAndGroup>();
                    if (_adapterListUserSelected != null)
                    {
                        _lstResult = _adapterListUserSelected.GetListIsclicked();
                    }
                    if (_lstResult == null) _lstResult = new List<BeanUserAndGroup>();
                    string _result = JsonConvert.SerializeObject(_lstResult);

                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, _result);
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, _result);
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                    }
                };
                #endregion

                _edtSearchChooseUser.Text = "";
                _edtSearchChooseUser.RequestFocus();

                if (elementParent.Enable == true) // Enabled = true mới cho clik
                {
                    CmmDroidFunction.ShowSoftKeyBoard(_rootView, _mainAct);
                }
            }
            catch (Exception ex)
            {
#if DEBUG
                CmmDroidFunction.WriteTrackingError(this.GetType().Name, "InitializeView", ex);
#endif
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/shareview/SharedView_PopupControlSelectUserGroupMulti.java
package com.vuthao.bpmop.shareview;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.custom.CustomFlexBoxRecyclerView;
import com.vuthao.bpmop.base.custom.flexbox.FlexDirection;
import com.vuthao.bpmop.base.custom.flexbox.FlexboxLayoutManager;
import com.vuthao.bpmop.base.custom.flexbox.JustifyContent;
import com.vuthao.bpmop.base.keyboard.KeyboardManager;
import com.vuthao.bpmop.base.model.app.Group;
import com.vuthao.bpmop.base.model.app.User;
import com.vuthao.bpmop.base.model.custom.UserAndGroup;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.realm.RealmController;
import com.vuthao.bpmop.core.Vars;
import com.vuthao.bpmop.detail.adapter.SelectUserGroupMultipleAdapter;
import com.vuthao.bpmop.shareview.adapter.SelectUserGroupMultiple_TextAdapter;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.stream.Collectors;

import io.realm.Realm;
import io.realm.RealmResults;

public class SharedView_PopupControlSelectUserGroupMulti extends SharedView_PopupControlBase implements TextWatcher, View.OnClickListener, SelectUserGroupMultiple_TextAdapter.SelectUserGroupMultiple_TextListener, SelectUserGroupMultipleAdapter.SelectUserGroupMultipleListener {
    private final boolean isUserAndGroup;
    private ArrayList<UserAndGroup> lstUserAndGroupAll = new ArrayList<UserAndGroup>();
    private ArrayList<UserAndGroup> lstSelected = new ArrayList<UserAndGroup>();
    private ArrayList<UserAndGroup> lstResult = new ArrayList<UserAndGroup>();
    private SelectUserGroupMultipleAdapter adapterListUser;
    private SelectUserGroupMultiple_TextAdapter adapterListUserSelected;
    private String result;
    private final Realm realm;

    private Dialog dialog;
    private View view;
    private LinearLayout lnSearch;
    private ImageView imgCloseChooseUser;
    private ImageView imgAcceptChooseUser;
    private ImageView imgDeleteChooseUser;
    private TextView tvTitleChooseUser;
    private EditText edtSearchChooseUser;
    private RecyclerView recyChooseUser;
    private CustomFlexBoxRecyclerView recySelectedUser;

    public SharedView_PopupControlSelectUserGroupMulti(LayoutInflater inflater, Activity mainAct, String fragmentTag, View rootView, boolean isUserAndGroup) {
        super(inflater, mainAct, fragmentTag, rootView);
        this.isUserAndGroup = isUserAndGroup;
        realm = new RealmController().getRealm();
    }

    @Override
    public void initializeValue_Master(ViewElement viewElement) {
        super.initializeValue_Master(viewElement);
    }

    @Override
    public void initializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JSONObject JObjectChild) {
        super.initializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
    }

    @Override
    public void initializeView() {
        super.initializeView();

        view = inflater.inflate(R.layout.popup_control_choose_user, null);
        lnSearch = view.findViewById(R.id.ln_PopupControl_ChooseUser_Search);
        imgCloseChooseUser = view.findViewById(R.id.img_PopupControl_ChooseUser_Close);
        imgAcceptChooseUser = view.findViewById(R.id.img_PopupControl_ChooseUser_Accept);
        imgDeleteChooseUser = view.findViewById(R.id.img_PopupControl_ChooseUser_Delete);
        tvTitleChooseUser = view.findViewById(R.id.tv_PopupControl_ChooseUser_Title);
        edtSearchChooseUser = view.findViewById(R.id.edt_PopupControl_ChooseUser_Search);
        recyChooseUser = view.findViewById(R.id.recy_PopupControl_ChooseUser);
        recySelectedUser = view.findViewById(R.id.recy_PopupControl_SelectedUser);

        recySelectedUser.setVisibility(View.VISIBLE);

        if (isUserAndGroup) {
            edtSearchChooseUser.setHint(Functions.share.getTitle("MESS_REQUIRE_USERGROUP", "Vui lòng chọn người hoặc nhóm để thực hiện"));
        } else {
            edtSearchChooseUser.setHint(Functions.share.getTitle("MESS_REQUIRE_USER", "Vui lòng chọn người thực hiện"));
        }

        boolean isEnable;
        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                tvTitleChooseUser.setText(elementParent.getTitle());
                if (!Functions.isNullOrEmpty(elementParent.getValue())) {
                    lstSelected = new Gson().fromJson(elementParent.getValue(), new TypeToken<ArrayList<UserAndGroup>>() {
                    }.getType());
                }

                isEnable = elementParent.isEnable();
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                tvTitleChooseUser.setText(elementPopup.getTitle());
                if (!Functions.isNullOrEmpty(elementPopup.getValue())) {
                    lstSelected = new Gson().fromJson(elementPopup.getValue(), new TypeToken<ArrayList<UserAndGroup>>() {
                    }.getType());
                }

                isEnable = elementPopup.isEnable();
                break;
            }
        }

        if (isEnable) {
            lnSearch.setVisibility(View.VISIBLE);
            imgDeleteChooseUser.setVisibility(View.VISIBLE);
            imgAcceptChooseUser.setVisibility(View.VISIBLE);
            recySelectedUser.setMaxRowAndRowHeight(Functions.share.convertDpToPixel(35f, mainAct.getBaseContext()), 3);

            lstUserAndGroupAll = getUserAndGroups();
            if (!lstUserAndGroupAll.isEmpty() && !lstSelected.isEmpty()) {
                for (int i = 0; i < lstSelected.size(); i++) {
                    int finalI = i;
                    lstUserAndGroupAll = (ArrayList<UserAndGroup>) lstUserAndGroupAll.stream().filter(r -> !r.getID().equals(lstSelected.get(finalI).getID())).collect(Collectors.toList());
                }
            }

            adapterListUser = new SelectUserGroupMultipleAdapter(mainAct, lstUserAndGroupAll, this);
            StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL);
            recyChooseUser.setAdapter(adapterListUser);
            recyChooseUser.setLayoutManager(staggeredGridLayoutManager);

            adapterListUserSelected = new SelectUserGroupMultiple_TextAdapter(mainAct, lstSelected, elementParent.isEnable(), this);
            FlexboxLayoutManager flexboxLayoutManager = new FlexboxLayoutManager(mainAct);
            flexboxLayoutManager.setFlexDirection(FlexDirection.ROW);
            flexboxLayoutManager.setJustifyContent(JustifyContent.FLEX_START);
            recySelectedUser.setAdapter(adapterListUserSelected);
            recySelectedUser.setLayoutManager(flexboxLayoutManager);

            edtSearchChooseUser.addTextChangedListener(this);
        } else {
            lnSearch.setVisibility(View.GONE);
            imgDeleteChooseUser.setVisibility(View.GONE);
            imgAcceptChooseUser.setVisibility(View.INVISIBLE);

            //trừ đi top navigation và / 35 số cứng
            int maxLines = (mainAct.getResources().getDisplayMetrics().heightPixels - (int) Functions.share.convertDpToPixel(50, mainAct.getBaseContext()) / Functions.share.convertDpToPixel(35, mainAct.getBaseContext()));
            recySelectedUser.setMaxRowAndRowHeight(Functions.share.convertDpToPixel(35, mainAct.getBaseContext()), maxLines);

            adapterListUserSelected = new SelectUserGroupMultiple_TextAdapter(mainAct, lstSelected, elementParent.isEnable(), pos -> {
            });

            FlexboxLayoutManager flexboxLayoutManager = new FlexboxLayoutManager(mainAct);
            flexboxLayoutManager.setFlexDirection(FlexDirection.ROW);
            flexboxLayoutManager.setJustifyContent(JustifyContent.FLEX_START);
            recySelectedUser.setAdapter(adapterListUserSelected);
            recySelectedUser.setLayoutManager(flexboxLayoutManager);
        }

        // region Dialog
        dialog = new Dialog(mainAct, R.style.Theme_Custom_BPMOP_Dialog_FullScreen);
        Window window = dialog.getWindow();
        dialog.requestWindowFeature(1);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setCancelable(true);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        window.setGravity(Gravity.BOTTOM);
        dialog.setContentView(view);
        dialog.show();

        WindowManager.LayoutParams s = window.getAttributes();
        s.width = WindowManager.LayoutParams.MATCH_PARENT;
        s.height = WindowManager.LayoutParams.MATCH_PARENT;
        window.setAttributes(s);
        // endregion

        imgCloseChooseUser.setOnClickListener(this);

        imgDeleteChooseUser.setOnClickListener(this);

        imgAcceptChooseUser.setOnClickListener(this);

        edtSearchChooseUser.setText("");
        edtSearchChooseUser.requestFocus();

        if (elementParent.isEnable()) {
            KeyboardManager.showKeyBoard(edtSearchChooseUser, mainAct);
        }
    }

    private void delete() {
        KeyboardManager.hideKeyboard(edtSearchChooseUser, mainAct);
        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow: {
                Intent intent = new Intent();
                intent.setAction("UPDATEFORM");
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", "");
                BroadcastUtility.send(mainAct, intent);
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                Intent intent = new Intent();
                intent.setAction("UPDATECHILDACTION");
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", "");
                BroadcastUtility.send(mainAct, intent);
                break;
            }
        }

        dialog.dismiss();
    }

    private void accept() {
        KeyboardManager.hideKeyboard(edtSearchChooseUser, mainAct);
        if (adapterListUserSelected != null) {
            lstResult = adapterListUserSelected.getListIsclicked();
        }

        result = new Gson().toJson(lstResult);
        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow: {
                Intent intent = new Intent();
                intent.setAction("UPDATEFORM");
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                Intent intent = new Intent();
                intent.setAction("UPDATECHILDACTION");
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                break;
            }
        }

        dialog.dismiss();
    }

    private ArrayList<UserAndGroup> getUserAndGroups() {
        ArrayList<UserAndGroup> _lstUserAndGroupAll = new ArrayList<>();
        if (isUserAndGroup) {
            RealmResults<User> results = realm.where(User.class).isNotNull("Email").findAll();
            for (User r : results) {
                UserAndGroup user = new UserAndGroup();
                user.setID(r.getID());
                user.setAccountName(r.getAccountName());
                user.setEmail(r.getEmail());
                user.setName(r.getFullName());
                user.setType("0");
                user.setImagePath(r.getImagePath());
                user.setSearch(r.getFullName() + r.getEmail());
                _lstUserAndGroupAll.add(user);
            }

            RealmResults<Group> groupRealmResults = realm.where(Group.class).isNotNull("Description").findAll();
            for (Group r : groupRealmResults) {
                UserAndGroup user = new UserAndGroup();
                user.setID(r.getID());
                user.setAccountName(r.getTitle());
                user.setEmail(r.getDescription());
                user.setName(r.getTitle());
                user.setType("1");
                user.setImagePath(r.getImage());
                user.setSearch(r.getTitle() + r.getDescription());
                _lstUserAndGroupAll.add(user);
            }
        } else {
            RealmResults<User> results = realm.where(User.class).isNotNull("Email").findAll();
            for (User r : results) {
                UserAndGroup user = new UserAndGroup();
                user.setID(r.getID());
                user.setAccountName(r.getAccountName());
                user.setEmail(r.getEmail());
                user.setName(r.getFullName());
                user.setType("0");
                user.setImagePath(r.getImagePath());
                user.setSearch(r.getFullName() + r.getEmail());
                _lstUserAndGroupAll.add(user);
            }
        }
        return _lstUserAndGroupAll;
    }

    @Override
    public void OnClick(UserAndGroup userAndGroup) {
        lstSelected.add(userAndGroup);
        lstUserAndGroupAll.remove(userAndGroup);
        adapterListUser.updateCurrentList(lstUserAndGroupAll);
        adapterListUserSelected.updateItemListIsClicked(lstSelected);
        adapterListUserSelected.notifyDataSetChanged();
        edtSearchChooseUser.setText("");
    }

    @Override
    public void OnDeleteItem(int pos) {
        lstUserAndGroupAll.add(lstSelected.get(pos));
        lstSelected.remove(lstSelected.get(pos));
        adapterListUser.updateCurrentList(lstUserAndGroupAll);
        adapterListUserSelected.updateItemListIsClicked(lstSelected);
        adapterListUserSelected.notifyDataSetChanged();
        edtSearchChooseUser.setText(edtSearchChooseUser.getText());
        edtSearchChooseUser.setSelection(edtSearchChooseUser.length());
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void afterTextChanged(Editable s) {
        adapterListUser.getFilter().filter(s.toString());
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.img_PopupControl_ChooseUser_Close: {
                KeyboardManager.hideKeyboard(edtSearchChooseUser, mainAct);
                dialog.dismiss();
                break;
            }
            case R.id.img_PopupControl_ChooseUser_Delete: {
                delete();
                break;
            }
            case R.id.img_PopupControl_ChooseUser_Accept: {
                accept();
                break;
            }
        }
    }
}
